File: fontes/frontend/fenix_provider/lib/src/model/nfe/nfe_cupom_fiscal_referenciado.dart
/*
Title: T2Ti ERP 3.0                                                                
Description: Model relacionado à tabela [NFE_CUPOM_FISCAL_REFERENCIADO] 
                                                                                
The MIT License                                                                 
                                                                                
Copyright: Copyright (C) 2021 T2Ti.COM                                          
                                                                                
Permission is hereby granted, free of charge, to any person                     
obtaining a copy of this software and associated documentation                  
files (the "Software"), to deal in the Software without                         
restriction, including without limitation the rights to use,                    
copy, modify, merge, publish, distribute, sublicense, and/or sell               
copies of the Software, and to permit persons to whom the                       
Software is furnished to do so, subject to the following                        
conditions:                                                                     
                                                                                
The above copyright notice and this permission notice shall be                  
included in all copies or substantial portions of the Software.                 
                                                                                
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,                 
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES                 
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND                        
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT                     
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,                    
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING                    
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR                   
OTHER DEALINGS IN THE SOFTWARE.                                                 
                                                                                
       The author may be contacted at:                                          
           t2ti.com@gmail.com                                                   
                                                                                
@author Albert Eije (alberteije@gmail.com)                    
@version 1.0.0
*******************************************************************************/
import 'dart:convert';

import 'package:intl/intl.dart';

class NfeCupomFiscalReferenciado {
	int? id;
	int? idNfeCabecalho;
	String? modeloDocumentoFiscal;
	int? numeroOrdemEcf;
	int? coo;
	DateTime? dataEmissaoCupom;
	int? numeroCaixa;
	String? numeroSerieEcf;

	NfeCupomFiscalReferenciado({
		id,
		idNfeCabecalho,
		modeloDocumentoFiscal,
		numeroOrdemEcf,
		coo,
		dataEmissaoCupom,
		numeroCaixa,
		numeroSerieEcf,
	});

	static List<String> campos = <String>[
		'ID', 
		'MODELO_DOCUMENTO_FISCAL', 
		'NUMERO_ORDEM_ECF', 
		'COO', 
		'DATA_EMISSAO_CUPOM', 
		'NUMERO_CAIXA', 
		'NUMERO_SERIE_ECF', 
	];
	
	static List<String> colunas = <String>[
		'Id', 
		'Modelo Documento Fiscal', 
		'Número de Ordem ECF', 
		'COO', 
		'Data de Emissão', 
		'Número Caixa', 
		'Número Série ECF', 
	];

	NfeCupomFiscalReferenciado.fromJson(Map<String, dynamic> jsonDados) {
		id = jsonDados['id'];
		idNfeCabecalho = jsonDados['idNfeCabecalho'];
		modeloDocumentoFiscal = getModeloDocumentoFiscal(jsonDados['modeloDocumentoFiscal']);
		numeroOrdemEcf = jsonDados['numeroOrdemEcf'];
		coo = jsonDados['coo'];
		dataEmissaoCupom = jsonDados['dataEmissaoCupom'] != null ? DateTime.tryParse(jsonDados['dataEmissaoCupom']) : null;
		numeroCaixa = jsonDados['numeroCaixa'];
		numeroSerieEcf = jsonDados['numeroSerieEcf'];
	}

	Map<String, dynamic> get toJson {
		Map<String, dynamic> jsonDados = <String, dynamic>{};

		jsonDados['id'] = id ?? 0;
		jsonDados['idNfeCabecalho'] = idNfeCabecalho ?? 0;
		jsonDados['modeloDocumentoFiscal'] = setModeloDocumentoFiscal(modeloDocumentoFiscal);
		jsonDados['numeroOrdemEcf'] = numeroOrdemEcf ?? 0;
		jsonDados['coo'] = coo ?? 0;
		jsonDados['dataEmissaoCupom'] = dataEmissaoCupom != null ? DateFormat('yyyy-MM-ddT00:00:00').format(dataEmissaoCupom!) : null;
		jsonDados['numeroCaixa'] = numeroCaixa ?? 0;
		jsonDados['numeroSerieEcf'] = numeroSerieEcf;
	
		return jsonDados;
	}
	
    getModeloDocumentoFiscal(String? modeloDocumentoFiscal) {
    	switch (modeloDocumentoFiscal) {
    		case '0':
    			return '2B=Cupom Fiscal emitido por máquina registradora';
    		case '1':
    			return '2C=Cupom Fiscal PDV';
    		case '2':
    			return '2D=Cupom Fiscal-emitido por ECF';
    		default:
    			return null;
    		}
    	}

    setModeloDocumentoFiscal(String? modeloDocumentoFiscal) {
    	switch (modeloDocumentoFiscal) {
    		case '2B=Cupom Fiscal emitido por máquina registradora':
    			return '0';
    		case '2C=Cupom Fiscal PDV':
    			return '1';
    		case '2D=Cupom Fiscal-emitido por ECF':
    			return '2';
    		default:
    			return null;
    		}
    	}


	String objetoEncodeJson(NfeCupomFiscalReferenciado objeto) {
	  final jsonDados = objeto.toJson;
	  return json.encode(jsonDados);
	}
	
}

File: fontes/backend/php/fenix/src/model/nfe/NfeCupomFiscalReferenciado.php
<?php
/*******************************************************************************
Title: T2Ti ERP 3.0
Description: Model relacionado à tabela [NFE_CUPOM_FISCAL_REFERENCIADO] 
                                                                                
The MIT License                                                                 
                                                                                
Copyright: Copyright (C) 2021 T2Ti.COM                                          
                                                                                
Permission is hereby granted, free of charge, to any person                     
obtaining a copy of this software and associated documentation                  
files (the "Software"), to deal in the Software without                         
restriction, including without limitation the rights to use,                    
copy, modify, merge, publish, distribute, sublicense, and/or sell               
copies of the Software, and to permit persons to whom the                       
Software is furnished to do so, subject to the following                        
conditions:                                                                     
                                                                                
The above copyright notice and this permission notice shall be                  
included in all copies or substantial portions of the Software.                 
                                                                                
THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,                 
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES                 
OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND                        
NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT                     
HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY,                    
WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING                    
FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR                   
OTHER DEALINGS IN THE SOFTWARE.                                                 
                                                                                
       The author may be contacted at:                                          
           t2ti.com@gmail.com                                                   
                                                                                
@author Albert Eije (alberteije@gmail.com)                    
@version 1.0.0
*******************************************************************************/
declare(strict_types=1);

class NfeCupomFiscalReferenciado extends EloquentModel implements \JsonSerializable
{
    /**
     * The table associated with the model.
     *
     * @var string
     */
    protected $table = 'NFE_CUPOM_FISCAL_REFERENCIADO';

    /**
     * Eager Loading - Relacionamentos que devem ser sempre carregados por padrão
     *
     * @var array
     */
	
    /**
     * Relations
     */
    public function nfeCabecalho()
    {
    	return $this->belongsTo(NfeCabecalho::class, 'ID_NFE_CABECALHO', 'ID');
    }


    /**
     * Gets e Sets
     */
    public function getIdAttribute() 
	{
		return $this->attributes['ID'];
	}

	public function setIdAttribute($id) 
	{
		$this->attributes['ID'] = $id;
	}

    public function getModeloDocumentoFiscalAttribute() 
	{
		return $this->attributes['MODELO_DOCUMENTO_FISCAL'];
	}

	public function setModeloDocumentoFiscalAttribute($modeloDocumentoFiscal) 
	{
		$this->attributes['MODELO_DOCUMENTO_FISCAL'] = $modeloDocumentoFiscal;
	}

    public function getNumeroOrdemEcfAttribute() 
	{
		return $this->attributes['NUMERO_ORDEM_ECF'];
	}

	public function setNumeroOrdemEcfAttribute($numeroOrdemEcf) 
	{
		$this->attributes['NUMERO_ORDEM_ECF'] = $numeroOrdemEcf;
	}

    public function getCooAttribute() 
	{
		return $this->attributes['COO'];
	}

	public function setCooAttribute($coo) 
	{
		$this->attributes['COO'] = $coo;
	}

    public function getDataEmissaoCupomAttribute() 
	{
		return $this->attributes['DATA_EMISSAO_CUPOM'];
	}

	public function setDataEmissaoCupomAttribute($dataEmissaoCupom) 
	{
		$this->attributes['DATA_EMISSAO_CUPOM'] = $dataEmissaoCupom;
	}

    public function getNumeroCaixaAttribute() 
	{
		return $this->attributes['NUMERO_CAIXA'];
	}

	public function setNumeroCaixaAttribute($numeroCaixa) 
	{
		$this->attributes['NUMERO_CAIXA'] = $numeroCaixa;
	}

    public function getNumeroSerieEcfAttribute() 
	{
		return $this->attributes['NUMERO_SERIE_ECF'];
	}

	public function setNumeroSerieEcfAttribute($numeroSerieEcf) 
	{
		$this->attributes['NUMERO_SERIE_ECF'] = $numeroSerieEcf;
	}


    /**
     * Mapping
     */
    public function mapear($objeto)
    {
        if (isset($objeto)) {
            isset($objeto->id) ? $this->setIdAttribute($objeto->id) : $this->setIdAttribute(null);

			$this->setModeloDocumentoFiscalAttribute($objeto->modeloDocumentoFiscal);
			$this->setNumeroOrdemEcfAttribute($objeto->numeroOrdemEcf);
			$this->setCooAttribute($objeto->coo);
			$this->setDataEmissaoCupomAttribute($objeto->dataEmissaoCupom);
			$this->setNumeroCaixaAttribute($objeto->numeroCaixa);
			$this->setNumeroSerieEcfAttribute($objeto->numeroSerieEcf);

			// vincular objetos
        }
    }

    /**
     * Serialization
     * {@inheritdoc}
     */
    public function jsonSerialize()
    {
        return [
			'id' => $this->getIdAttribute(),
			'modeloDocumentoFiscal' => $this->getModeloDocumentoFiscalAttribute(),
			'numeroOrdemEcf' => $this->getNumeroOrdemEcfAttribute(),
			'coo' => $this->getCooAttribute(),
			'dataEmissaoCupom' => $this->getDataEmissaoCupomAttribute(),
			'numeroCaixa' => $this->getNumeroCaixaAttribute(),
			'numeroSerieEcf' => $this->getNumeroSerieEcfAttribute(),
        ];
    }
}